File: clients/dotnet/src/Varta.Client/Internal/VlpSecure/AeadCodec.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace Varta.Internal.VlpSecure;

/// <summary>
/// ChaCha20-Poly1305 wire codec for the Varta secure UDP transport.
/// Two modes:
/// <list type="bullet">
///   <item><c>Shared</c> — 32-byte pre-shared key, 60-byte wire frame.</item>
///   <item><c>Master</c> — 32-byte master key + per-agent key derived
///     via HKDF (agent_pid as info); 64-byte wire frame with agent_pid
///     as AAD.</item>
/// </list>
/// Spec: <c>book/src/spec/vlp-secure.md</c>. Conformance vectors:
/// <c>tools/vlp-test-vectors.json</c> under <c>secure_frame_vectors</c>.
/// </summary>
internal static class AeadCodec
{
    public const int KeyBytes = 32;
    public const int IvRandomBytes = 8;
    public const int IvCounterBytes = 4;
    public const int AeadNonceBytes = 12;
    public const int TagBytes = 16;
    public const int PlaintextBytes = 32;
    public const int SharedFrameBytes = 60; // iv_random[8] + iv_counter[4] + ct[32] + tag[16]
    public const int MasterFrameBytes = 64; // agent_pid[4] + iv_random[8] + iv_counter[4] + ct[32] + tag[16]
    public const int AgentPidBytes = 4;

    public static void EncodeShared(
        ReadOnlySpan<byte> key32,
        ReadOnlySpan<byte> ivRandom8,
        uint ivCounter,
        ReadOnlySpan<byte> plaintext32,
        Span<byte> destination60)
    {
        if (key32.Length != KeyBytes) throw new ArgumentException("key must be 32 bytes", nameof(key32));
        if (ivRandom8.Length != IvRandomBytes) throw new ArgumentException("ivRandom must be 8 bytes", nameof(ivRandom8));
        if (plaintext32.Length != PlaintextBytes) throw new ArgumentException("plaintext must be 32 bytes", nameof(plaintext32));
        if (destination60.Length != SharedFrameBytes) throw new ArgumentException($"destination must be {SharedFrameBytes} bytes", nameof(destination60));

        Span<byte> nonce = stackalloc byte[AeadNonceBytes];
        ivRandom8.CopyTo(nonce);
        BinaryPrimitives.WriteUInt32LittleEndian(nonce[IvRandomBytes..], ivCounter);

        ivRandom8.CopyTo(destination60[..IvRandomBytes]);
        BinaryPrimitives.WriteUInt32LittleEndian(destination60.Slice(IvRandomBytes, IvCounterBytes), ivCounter);

        Span<byte> ciphertext = destination60.Slice(IvRandomBytes + IvCounterBytes, PlaintextBytes);
        Span<byte> tag = destination60.Slice(IvRandomBytes + IvCounterBytes + PlaintextBytes, TagBytes);

        using var aead = new ChaCha20Poly1305(key32);
        aead.Encrypt(nonce, plaintext32, ciphertext, tag, associatedData: ReadOnlySpan<byte>.Empty);
    }

    public static void DecodeShared(ReadOnlySpan<byte> key32, ReadOnlySpan<byte> wire60, Span<byte> plaintext32)
    {
        if (key32.Length != KeyBytes) throw new ArgumentException("key must be 32 bytes", nameof(key32));
        if (wire60.Length != SharedFrameBytes) throw new ArgumentException($"wire must be {SharedFrameBytes} bytes", nameof(wire60));
        if (plaintext32.Length != PlaintextBytes) throw new ArgumentException($"plaintext must be {PlaintextBytes} bytes", nameof(plaintext32));

        ReadOnlySpan<byte> nonce = wire60[..AeadNonceBytes]; // iv_random[8] || iv_counter[4]
        ReadOnlySpan<byte> ciphertext = wire60.Slice(AeadNonceBytes, PlaintextBytes);
        ReadOnlySpan<byte> tag = wire60.Slice(AeadNonceBytes + PlaintextBytes, TagBytes);

        using var aead = new ChaCha20Poly1305(key32);
        aead.Decrypt(nonce, ciphertext, tag, plaintext32, associatedData: ReadOnlySpan<byte>.Empty);
    }

    public static void EncodeMaster(
        ReadOnlySpan<byte> masterKey32,
        uint agentPid,
        ReadOnlySpan<byte> ivRandom8,
        uint ivCounter,
        ReadOnlySpan<byte> plaintext32,
        Span<byte> destination64,
        Span<byte> derivedAgentKey32)
    {
        if (masterKey32.Length != KeyBytes) throw new ArgumentException("masterKey must be 32 bytes", nameof(masterKey32));
        if (ivRandom8.Length != IvRandomBytes) throw new ArgumentException("ivRandom must be 8 bytes", nameof(ivRandom8));
        if (plaintext32.Length != PlaintextBytes) throw new ArgumentException("plaintext must be 32 bytes", nameof(plaintext32));
        if (destination64.Length != MasterFrameBytes) throw new ArgumentException($"destination must be {MasterFrameBytes} bytes", nameof(destination64));
        if (derivedAgentKey32.Length != KeyBytes) throw new ArgumentException("derivedAgentKey must be 32 bytes", nameof(derivedAgentKey32));

        Hkdf.DeriveAgentKey(masterKey32, agentPid, derivedAgentKey32);

        BinaryPrimitives.WriteUInt32LittleEndian(destination64[..AgentPidBytes], agentPid);
        ivRandom8.CopyTo(destination64.Slice(AgentPidBytes, IvRandomBytes));
        BinaryPrimitives.WriteUInt32LittleEndian(destination64.Slice(AgentPidBytes + IvRandomBytes, IvCounterBytes), ivCounter);

        Span<byte> nonce = stackalloc byte[AeadNonceBytes];
        ivRandom8.CopyTo(nonce);
        BinaryPrimitives.WriteUInt32LittleEndian(nonce[IvRandomBytes..], ivCounter);

        Span<byte> ciphertext = destination64.Slice(AgentPidBytes + AeadNonceBytes, PlaintextBytes);
        Span<byte> tag = destination64.Slice(AgentPidBytes + AeadNonceBytes + PlaintextBytes, TagBytes);
        ReadOnlySpan<byte> aad = destination64[..AgentPidBytes];

        using var aead = new ChaCha20Poly1305(derivedAgentKey32);
        aead.Encrypt(nonce, plaintext32, ciphertext, tag, associatedData: aad);
    }

    public static void DecodeMaster(ReadOnlySpan<byte> masterKey32, ReadOnlySpan<byte> wire64, Span<byte> plaintext32)
    {
        if (masterKey32.Length != KeyBytes) throw new ArgumentException("masterKey must be 32 bytes", nameof(masterKey32));
        if (wire64.Length != MasterFrameBytes) throw new ArgumentException($"wire must be {MasterFrameBytes} bytes", nameof(wire64));
        if (plaintext32.Length != PlaintextBytes) throw new ArgumentException($"plaintext must be {PlaintextBytes} bytes", nameof(plaintext32));

        uint agentPid = BinaryPrimitives.ReadUInt32LittleEndian(wire64[..AgentPidBytes]);
        Span<byte> agentKey = stackalloc byte[KeyBytes];
        Hkdf.DeriveAgentKey(masterKey32, agentPid, agentKey);

        ReadOnlySpan<byte> nonce = wire64.Slice(AgentPidBytes, AeadNonceBytes);
        ReadOnlySpan<byte> ciphertext = wire64.Slice(AgentPidBytes + AeadNonceBytes, PlaintextBytes);
        ReadOnlySpan<byte> tag = wire64.Slice(AgentPidBytes + AeadNonceBytes + PlaintextBytes, TagBytes);
        ReadOnlySpan<byte> aad = wire64[..AgentPidBytes];

        using var aead = new ChaCha20Poly1305(agentKey);
        aead.Decrypt(nonce, ciphertext, tag, plaintext32, associatedData: aad);
    }
}

File: clients/java/lib/src/main/java/health/varta/vlpsecure/AeadCodec.java
package health.varta.vlpsecure;

import javax.crypto.AEADBadTagException;
import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.security.GeneralSecurityException;
import java.util.Objects;

/**
 * ChaCha20-Poly1305 AEAD seal/open for VLP-Secure frames.
 *
 * <p>Two modes:
 * <ul>
 *   <li><b>Shared</b> — 60-byte wire: {@code iv_random[8] || iv_counter[4] || ct[32] || tag[16]}.
 *       AAD = empty.</li>
 *   <li><b>Master</b> — 64-byte wire: {@code agent_pid[4] || iv_random[8] || iv_counter[4] || ct[32] || tag[16]}.
 *       AAD = {@code agent_pid} LE bytes (bound by Poly1305). Per-agent key derived via
 *       {@link Hkdf#deriveAgentKey(byte[], int)}.</li>
 * </ul>
 *
 * <p>Spec: {@code book/src/spec/vlp-secure.md} §3–§5.</p>
 */
public final class AeadCodec {
    private AeadCodec() {}

    /** Algorithm string for {@link Cipher#getInstance(String)}. */
    public static final String CIPHER_NAME = "ChaCha20-Poly1305";
    /** JCE key algorithm name for {@link SecretKeySpec}. */
    public static final String KEY_ALG = "ChaCha20";

    /**
     * Seal a 32-byte plaintext under a shared key. Output is exactly 60 bytes
     * laid out as {@code iv_random || iv_counter || ciphertext || tag}.
     */
    public static byte[] encodeShared(byte[] sharedKey32, byte[] ivRandom8, int ivCounter, byte[] plaintext32) {
        requireLen("sharedKey32", sharedKey32, Hkdf.KEY_BYTES);
        requireLen("ivRandom8", ivRandom8, Hkdf.IV_RANDOM_BYTES);
        requireLen("plaintext32", plaintext32, 32);

        byte[] nonce = buildNonce(ivRandom8, ivCounter);
        byte[] sealed = aeadSeal(sharedKey32, nonce, plaintext32, /* aad */ null);

        byte[] wire = new byte[Hkdf.SECURE_SHARED_BYTES];
        System.arraycopy(ivRandom8, 0, wire, 0, Hkdf.IV_RANDOM_BYTES);
        ByteBuffer.wrap(wire, Hkdf.IV_RANDOM_BYTES, Hkdf.IV_COUNTER_BYTES)
            .order(ByteOrder.LITTLE_ENDIAN).putInt(ivCounter);
        System.arraycopy(sealed, 0, wire, Hkdf.IV_RANDOM_BYTES + Hkdf.IV_COUNTER_BYTES, sealed.length);
        return wire;
    }

    /**
     * Open a 60-byte shared-key wire frame. Throws {@link AEADBadTagException}
     * (rethrown as {@link SecurityException}) on tag mismatch.
     */
    public static byte[] decodeShared(byte[] sharedKey32, byte[] wire60) {
        requireLen("sharedKey32", sharedKey32, Hkdf.KEY_BYTES);
        requireLen("wire60", wire60, Hkdf.SECURE_SHARED_BYTES);

        byte[] ivRandom = new byte[Hkdf.IV_RANDOM_BYTES];
        System.arraycopy(wire60, 0, ivRandom, 0, Hkdf.IV_RANDOM_BYTES);
        int ivCounter = ByteBuffer.wrap(wire60, Hkdf.IV_RANDOM_BYTES, Hkdf.IV_COUNTER_BYTES)
            .order(ByteOrder.LITTLE_ENDIAN).getInt();
        byte[] ctTag = new byte[32 + Hkdf.TAG_BYTES];
        System.arraycopy(wire60, Hkdf.IV_RANDOM_BYTES + Hkdf.IV_COUNTER_BYTES, ctTag, 0, ctTag.length);

        byte[] nonce = buildNonce(ivRandom, ivCounter);
        return aeadOpen(sharedKey32, nonce, ctTag, /* aad */ null);
    }

    /**
     * Seal a 32-byte plaintext under a master key for {@code agentPid}.
     * Output is exactly 64 bytes laid out as
     * {@code agent_pid || iv_random || iv_counter || ciphertext || tag}.
     * AAD = {@code u32_LE(agent_pid)} (4 bytes).
     */
    public static byte[] encodeMaster(byte[] masterKey32, int agentPid, byte[] ivRandom8, int ivCounter,
                                      byte[] plaintext32) {
        requireLen("masterKey32", masterKey32, Hkdf.KEY_BYTES);
        requireLen("ivRandom8", ivRandom8, Hkdf.IV_RANDOM_BYTES);
        requireLen("plaintext32", plaintext32, 32);

        byte[] agentKey = Hkdf.deriveAgentKey(masterKey32, agentPid);
        byte[] aad = new byte[Integer.BYTES];
        ByteBuffer.wrap(aad).order(ByteOrder.LITTLE_ENDIAN).putInt(agentPid);

        byte[] nonce = buildNonce(ivRandom8, ivCounter);
        byte[] sealed = aeadSeal(agentKey, nonce, plaintext32, aad);

        byte[] wire = new byte[Hkdf.SECURE_MASTER_BYTES];
        System.arraycopy(aad, 0, wire, 0, Integer.BYTES);
        System.arraycopy(ivRandom8, 0, wire, Integer.BYTES, Hkdf.IV_RANDOM_BYTES);
        ByteBuffer.wrap(wire, Integer.BYTES + Hkdf.IV_RANDOM_BYTES, Hkdf.IV_COUNTER_BYTES)
            .order(ByteOrder.LITTLE_ENDIAN).putInt(ivCounter);
        System.arraycopy(sealed, 0, wire,
            Integer.BYTES + Hkdf.IV_RANDOM_BYTES + Hkdf.IV_COUNTER_BYTES, sealed.length);

        java.util.Arrays.fill(agentKey, (byte) 0);
        return wire;
    }

    /**
     * Open a 64-byte master-key wire frame. Re-derives the per-agent key from
     * {@code agent_pid} on the wire.
     */
    public static byte[] decodeMaster(byte[] masterKey32, byte[] wire64) {
        requireLen("masterKey32", masterKey32, Hkdf.KEY_BYTES);
        requireLen("wire64", wire64, Hkdf.SECURE_MASTER_BYTES);

        int agentPid = ByteBuffer.wrap(wire64, 0, Integer.BYTES)
            .order(ByteOrder.LITTLE_ENDIAN).getInt();
        byte[] ivRandom = new byte[Hkdf.IV_RANDOM_BYTES];
        System.arraycopy(wire64, Integer.BYTES, ivRandom, 0, Hkdf.IV_RANDOM_BYTES);
        int ivCounter = ByteBuffer.wrap(wire64, Integer.BYTES + Hkdf.IV_RANDOM_BYTES, Hkdf.IV_COUNTER_BYTES)
            .order(ByteOrder.LITTLE_ENDIAN).getInt();
        byte[] ctTag = new byte[32 + Hkdf.TAG_BYTES];
        System.arraycopy(wire64, Integer.BYTES + Hkdf.IV_RANDOM_BYTES + Hkdf.IV_COUNTER_BYTES,
            ctTag, 0, ctTag.length);

        byte[] agentKey = Hkdf.deriveAgentKey(masterKey32, agentPid);
        byte[] aad = new byte[Integer.BYTES];
        ByteBuffer.wrap(aad).order(ByteOrder.LITTLE_ENDIAN).putInt(agentPid);
        byte[] nonce = buildNonce(ivRandom, ivCounter);
        try {
            return aeadOpen(agentKey, nonce, ctTag, aad);
        } finally {
            java.util.Arrays.fill(agentKey, (byte) 0);
        }
    }

    static byte[] buildNonce(byte[] ivRandom8, int ivCounter) {
        byte[] nonce = new byte[Hkdf.NONCE_BYTES];
        System.arraycopy(ivRandom8, 0, nonce, 0, Hkdf.IV_RANDOM_BYTES);
        ByteBuffer.wrap(nonce, Hkdf.IV_RANDOM_BYTES, Hkdf.IV_COUNTER_BYTES)
            .order(ByteOrder.LITTLE_ENDIAN).putInt(ivCounter);
        return nonce;
    }

    private static byte[] aeadSeal(byte[] key32, byte[] nonce12, byte[] plaintext, byte[] aadOrNull) {
        try {
            Cipher c = Cipher.getInstance(CIPHER_NAME);
            c.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key32, KEY_ALG), new IvParameterSpec(nonce12));
            if (aadOrNull != null) {
                c.updateAAD(aadOrNull);
            }
            return c.doFinal(plaintext);
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("ChaCha20-Poly1305 unavailable", e);
        }
    }

    private static byte[] aeadOpen(byte[] key32, byte[] nonce12, byte[] ciphertextAndTag, byte[] aadOrNull) {
        try {
            Cipher c = Cipher.getInstance(CIPHER_NAME);
            c.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key32, KEY_ALG), new IvParameterSpec(nonce12));
            if (aadOrNull != null) {
                c.updateAAD(aadOrNull);
            }
            return c.doFinal(ciphertextAndTag);
        } catch (AEADBadTagException e) {
            throw new SecurityException("AEAD tag mismatch", e);
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("ChaCha20-Poly1305 unavailable", e);
        }
    }

    private static void requireLen(String name, byte[] b, int expected) {
        Objects.requireNonNull(b, name);
        if (b.length != expected) {
            throw new IllegalArgumentException(name + " must be " + expected + " bytes, got " + b.length);
        }
    }
}
